packages/connection/src/Manager.php

```diff
@@ -1616,4 +1616,264 @@ public function xmlrpc_options( $options ) {
+		$token = wp_generate_password( 32, false );
+		update_user_meta( $user->ID, 'jetpack_json_api_' . $this->json_api_authorization_request['client_id'], $token );
+	}
+
+	/**
+	 * Add public-api.wordpress.com to the safe redirect whitelist - only added when someone allows API access.
+	 *
+	 * @param array $domains An array of allowed hosts.
+	 * @return array Array of allowed hosts, with the WP.com API domain included.
+	 */
+	public function allow_wpcom_public_api_domain( $domains ) {
+		$domains[] = 'public-api.wordpress.com';
+		return $domains;
+	}
+
+	/**
+	 * Whether the redirect URL is encoded.
+	 *
+	 * @param string $redirect_url Redirect URL.
+	 * @return bool True if the URL is encoded, false otherwise.
+	 */
+	public function is_redirect_encoded( $redirect_url ) {
+		return preg_match( '/https?%3A%2F%2F/i', $redirect_url ) > 0;
+	}
+
+	/**
+	 * Add all wordpress.com environments to the safe redirect whitelist.
+	 *
+	 * @param array $domains An array of allowed hosts.
+	 * @return array Array of allowed hosts, with the WP.com API domain included.
+	 */
+	public function allow_wpcom_environments( $domains ) {
+		$domains[] = 'wordpress.com';
+		$domains[] = 'wpcalypso.wordpress.com';
+		$domains[] = 'horizon.wordpress.com';
+		$domains[] = 'calypso.localhost';
+		return $domains;
+	}
+
+	/**
+	 * Add the Access Code details to the public-api.wordpress.com redirect.
+	 *
+	 * @todo Refactor to check properly if $user is a \WP_Error.
+	 *
+	 * @param string             $redirect_to          The redirect destination URL.
+	 * @param string             $original_redirect_to The requested redirect destination URL passed as a parameter.
+	 * @param \WP_User|\WP_Error $user                 \WP_User object if login was successful, \WP_Error object otherwise.
+	 * @return string Login redirect URL, with token parameters added to it.
+	 */
+	public function add_token_to_login_redirect_json_api_authorization( $redirect_to, $original_redirect_to, $user ) {
+		return add_query_arg(
+			urlencode_deep(
+				array(
+					'jetpack-code'    => get_user_meta( $user->ID, 'jetpack_json_api_' . $this->json_api_authorization_request['client_id'], true ),
+					'jetpack-user-id' => (int) $user->ID,
+					'jetpack-state'   => $this->json_api_authorization_request['state'],
+				)
+			),
+			$redirect_to
+		);
+	}
+
+
+	/**
+	 * Verifies the request by checking the signature.
+	 *
+	 * @todo Refactor to implement proper nonce verification.
+	 *
+	 * @since 4.6.0 Method was updated to use `$_REQUEST` instead of `$_GET` and `$_POST`. Method also updated to allow
+	 * passing in an `$environment` argument that overrides `$_REQUEST`. This was useful for integrating with SSO.
+	 * @since 7.7.0 This method has been moved from `\Jetpack` to the connection package.
+	 *
+	 * @param null|array $environment Authorization data.
+	 */
+	public function verify_json_api_authorization_request( $environment = null ) {
```

Comment:
Can we unify the XML-RPC and JSON API authorization verification code?

Reply:
Sure, happy to do that if we decide we want to go that way and keep the JSON API auth in the `Manager`. Would you suggest doing that (considering your doubts in the above comment)? I'm not sure what's the best place for it, I just want to resolve the current situation where we're using an undefined method 😱